presto-common/src/main/java/com/facebook/presto/common/Page.java

```diff
@@ -291,22 +291,23 @@ public Page compact()
      */
     public Page getLoadedPage()
     {
-        Block[] loadedBlocks = null;
         for (int i = 0; i < blocks.length; i++) {
             Block loaded = blocks[i].getLoadedBlock();
             if (loaded != blocks[i]) {
-                if (loadedBlocks == null) {
-                    loadedBlocks = blocks.clone();
+                // Transition to new block creation mode after the first newly loaded block is encountered
+                Block[] loadedBlocks = new Block[blocks.length];
+                if (i > 0) {
+                    System.arraycopy(blocks, 0, loadedBlocks, 0, i);
                 }
-                loadedBlocks[i] = loaded;
+                loadedBlocks[i++] = loaded;
+                for (; i < loadedBlocks.length; i++) {
+                    loadedBlocks[i] = blocks[i].getLoadedBlock();
+                }
```

Comment:
same question as the other one.

Reply:
See https://github.com/prestodb/presto/pull/16683#discussion_r702470537 for the rationale.